### PR TITLE
MC-39718: Process Manager always exits successfully

### LINKED ISSUE #30964: Process Manager always exits successfully if the amount of functions(i.e. indexer dimensions) is lower than the MAGE_INDEXER_THREADS_COUNT env variable
Process Manager always exits successfully if the amount of functions passed to it(i.e. indexer dimensions) is lower than the MAGE*INDEXER_THREADS*COUNT env variable. Can cause the Index to think it was rebuilt successfully when in fact it wasn't.
### Preconditions (*)

1. Magento 2.3.0 CE/EE or higher
1. Magento 2.4-develop
1. Several stores are active, but less than MAGE*INDEXER_THREADS_COUNT(if MAGE_INDEXER_THREADS*COUNT=4, then the amount of stores active to reproduce this bug can be 3).
1. (Optional) Elasticsearch is configured as search_engine - for ease of reproducing

### Steps to reproduce (*)

1. Run `MAGE*INDEXER_THREADS_COUNT=4 php bin/magento indexer:reindex catalogsearch*fulltext`
1. Kill Elasticsearch process to simulate an exception thrown

### Expected result (*)

1. Main process exits with an error as child process failed.

### Actual result (*)

1. Observe "Catalog Search index has been rebuilt successfully"
```
xx@xx:/var/www/magento2$ MAGE*INDEXER_THREADS_COUNT=4 php bin/magento indexer:reindex catalogsearch*fulltext
Catalog Search indexer process unknown error:
Indexer handler is not available: elasticsearch7
Catalog Search indexer process unknown error:
Indexer handler is not available: elasticsearch7
Catalog Search indexer process unknown error:
Indexer handler is not available: elasticsearch7
Catalog search index has been rebuilt successfully in ...
```
(Sidenote) The duplicated error messages are already fixed in scope of https://github.com/magento/magento2/pull/30626
## 
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- <x] Severity: **S0** *- Affects critical data or functionality and leaves users without workaround.*
- [ ] Severity: **S1** *- Affects critical data or functionality and forces users to employ a workaround.*
- [ ] Severity: **S2** *- Affects non-critical data or functionality and forces users to employ a workaround.*
- [ ] Severity: **S3** *- Affects non-critical data or functionality and does not force users to employ a workaround.*
- [ > Severity: **S4** *- Affects aesthetics, professional look and feel, “quality” or “usability”.*

### PR DESCRIPTION
**Pull request by engcom-Golf**
if the amount of functions(i.e. indexer dimensions) is lower than the MAGE_INDEXER_THREADS_COUNT env variable

### Description (*)

Added logic to check if child process has exit code >0 then mark parent process as failed
### Fixed Issues (if relevant)

1. Fixes magento/magento2#30964

### Manual testing scenarios (*)

1. Configure a second website with store and store view
2. Generate a small profile for performance testing
3. Run fulltext_search reindex with parameter MAGE_INDEXER_THREADS_COUNT > {stores count}
4. While reindex is not finished disable elasticsearch in a separate terminal window
#### Expected result:
The message "reindex successfully finished" is not shown
Reindex is not marked as valid in admin

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Golf** [comment]
@magento run all tests

File: app/code/Magento/Indexer/Model/ProcessManager.php [hunk 1]
```diff
@@ -114,6 +114,9 @@ private function multiThreadsExecute($userFunctions)
         // phpcs:ignore Magento2.CodeAnalysis.EmptyBlock,Magento2.Functions.DiscouragedFunction
```
Comment (ihor-sviziev) [inline app/code/Magento/Indexer/Model/ProcessManager.php]: I believe this ignoring of empty block could be removed?
```suggestion
        // phpcs:ignore Magento2.Functions.DiscouragedFunction
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Indexer/Test/Unit/Model/ProcessManagerTest.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * Closure functions data provider for multi thread execution
+     *
+     * @return array
+     * @SuppressWarnings(PHPMD.ExitExpression)
+     */
+    public function closureFunctionsProvider(): array
+    {
+        return [
+            'more_threads_than_functions' => [
```
Comment (ihor-sviziev) [inline app/code/Magento/Indexer/Test/Unit/Model/ProcessManagerTest.php:51]: Could you also add coverage for cases when it's mess threads than functions and when it's equal? Just to make sure it works fine as well

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Indexer/Test/Unit/Model/ProcessManagerTest.php (hunk 2 above)
Comment (engcom-Golf) [reply app/code/Magento/Indexer/Test/Unit/Model/ProcessManagerTest.php:51]: thanks for looking into it, will do

File: app/code/Magento/Indexer/Model/ProcessManager.php (hunk 1 above)
Comment (engcom-Golf) [reply app/code/Magento/Indexer/Model/ProcessManager.php]: updated

File: app/code/Magento/Indexer/Test/Unit/Model/ProcessManagerTest.php (hunk 2 above)
Comment (engcom-Golf) [reply app/code/Magento/Indexer/Test/Unit/Model/ProcessManagerTest.php:51]: added variations

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Review by engcom-Kilo — APPROVED** [review]

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8567](https://jira.corp.magento.com/browse/ENGCOM-8567) has been created to process this Pull Request

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA Passed

### Manual testing scenario
1. Generate a small profile for performance testing `bin/magento setup:performance:generate-fixtures setup/performance-toolkit/profiles/ce/small.xml`
2. Run fulltext_search reindex with parameter MAGE_INDEXER_THREADS_COUNT > {stores count} `MAGE_INDEXER_THREADS_COUNT=4 php bin/magento indexer:reindex catalogsearch_fulltext`
3. While reindex is not finished, disable Elasticsearch in a separate terminal window `sudo service elasticsearch stop`

**Before** applying changes provided in the PR, the Process Manager exited successfully
exitedpng

**After** switching to the PR, the Process Manager exits with _Fail in child process_ error
exitedpng_PR

**Comment by zakdma** [comment]
@magento import code to https://github.com/magento-tsg/magento2ce

**Comment by magento-engcom-team** [comment]
@zakdma the branch with code successfully imported into`magento-tsg/magento2ce` repository. Branch name: `imported-magento-magento2-31368`.

**Pull request merged**